R/Quosure.R

```diff
@@ -0,0 +1,270 @@
+#' - if their environments contains objects of the same name but not identical
+#' - if `object2` has unique code element placed before common element. This means that `object2`
+#' in the environment of the `object2` was evaluated some extra code before which can influence
+#' reproducibility
+#' - more cases todo
+#' @param object (`Quosure`)
+#' @param object2 (`Quosure`)
+#' @param overwrite (`logical(1)`) whether modified objects in the environment of the `object2`
+#'  should overwrite their equivalents in the environment of the `object`.
+#' @export
+setGeneric("join", function(object, object2, overwrite = FALSE) {
+  standardGeneric("join")
+})
+
+setMethod("join", signature("Quosure", "Quosure"), function(object, object2, overwrite = FALSE) {
+  # todo: revise chunks_merge_chunks and make robust join method
+
+  # object2 can't have modified object of the same name! See chunks_push_chunks
+  common_names <- intersect(ls(object@env), ls(object2@env))
+  is_identical_obj <- vapply(common_names, function(x) {
+    identical(get(x, object@env), get(x, object2@env))
+  }, logical(1))
+  if (any(isFALSE(is_identical_obj)) && !overwrite) {
+    stop(
+      "join does not allow overwriting already calculated values.",
+      " Following variables would have been overwritten:",
+      paste("    -", names(is_identical_obj), collapse = "\n"),
+      sep = "\n"
+    )
+  }
+
+  # join expressions
+  # > join should be an inteligent union
+  unique_expr <- !(object2@code %in% object@code & names(object2@code) %in% names(object@code))
+    # duplicated expr will probably have the same name also
+    # unique_expr should have TRUE only at the end of the vector.
+  object@code <- c(object@code, object2@code[unique_expr])
+
+  # insert new objects to env
+  new_names <- setdiff(ls(object2@env), ls(object@env))
+
+  object@env <- .copy_env(object@env)
+  lapply(new_names, function(x) assign(x, get(x, object2@env), object@env))
+
+  object
+})
+
+#' Copy environment content to the other environment
+#' To avoid sharing the same environment by multiplequo object (in reactives)
+#' @keywords internal
+.copy_env <- function(env) {
+  # need to copy the objects from old env to new env
+  # to avoid "chunks" sharing the same environment
+  env_new <- new.env()
+  for (i in ls(env)) {
+    env_new[[i]] <- env[[i]]
+  }
+  env_new
+}
+
+#' Needed to determine if the `Quosure`'s code can be joined
+#'
+#' @keywords internal
+.resolve_code_vector <- function(code, code2) {
+  # only possible pattern for extra code: both are completelly different
+  # then A <- B (but there should be no common variable names!)? what about rm method?
+
+  # if  B has extra code which is not  at the end
+  # only one object can have extra code at the
+
+  # if B has extra code at the end and no object conflict
+  # if A has extra code at the end and no object conflict
+  #
+
+  code_only_in_a <- which(!code %in% code2)
+  code_only_in_b <- which(!code2 %in% code)
```

Comment:
These feels risky - how about?

```
for ( i in 1:10) {
  x[[i]] <- 5
}
```

and

```
for ( i in 1:10) {
  print("Hello gogonzo")
}
```

It feels like we should be able to leverage shiny's reactivity handling here as that doesn't seem to have a problem knowing what depends on what??